### PR TITLE
rgw/multisite: revert "rgw/multisite: avoid redundant error repo entry logging

### PR DESCRIPTION
**Pull request by clwluvw**
Pass the requested gen than the bucket_status.incremental_gen to RGWSyncBucketShardCR as the old gen (bucket_status.incremental_gen) is stored in the error repo and will retried later.

Fixes: https://tracker.ceph.com/issues/68805

### CODE REVIEW

File: src/rgw/driver/rados/rgw_data_sync.cc [hunk 1]
```diff
@@ -6079,7 +6079,7 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
 
         yield call(new RGWSyncBucketShardCR(sc, data_lease_cr, sync_pair,
                                             sync_pipe, bucket_stopped,
-                                            bucket_status.incremental_gen, tn, progress));
+                                            gen ? *gen : bucket_status.incremental_gen, tn, progress));
```
[the remark this answers is not in the document]
Comment (clwluvw) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: @cbodley it's not - actually with this commit (https://github.com/ceph/ceph/commit/a191f9a6d2997f655f580e16731615e77350a03d) it will continue as there is an `else` surrounding it.

[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: thanks, that looks like a bug? we probably want a `return set_cr_done()` in the first branch. cc @smanjara

[the remark this answers is not in the document]
Comment (clwluvw) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: that was my confusion, but if we return cr done what would happen to the requested gen? how is that gonna retry later? are we logging the requested gen anywhere?

[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: true, that branch is adding an error repo entry for the current_gen, not gen. maybe the error repo entry wasn't redundant after all?

[the remark this answers is not in the document]
Comment (clwluvw) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: perhaps, but why shouldn't we continue with the requested gen? any possible problems?

[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: it's a matter of prioritization. we want to focus on finishing the oldest generation first so we can retire it's logs

[the remark this answers is not in the document]
Comment (smanjara) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: oh probably it was retrying requested gen that I thought was redundant here https://github.com/ceph/ceph/pull/59580?

[the remark this answers is not in the document]
Comment (clwluvw) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: probably, but still, I'm confused as to why the CR can't continue with the request gen and we should log both in the error repo. why can't we utilize the current CR?

[the remark this answers is not in the document]
Comment (smanjara) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: the idea is to finish processing whatever generation we get from  `bucket_status.incremental_gen` first so that we mark the shards for that generation as done before we move on to the next generation in an order during bucket incremental sync. (its contrary to what I said in the chat. apologies for that) we update the shards as done with a particular generation in the bucket sync status [here](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L5187) 
so any requests that comes with a different gen should go to the error repo which would correctly happen if we reverted https://github.com/ceph/ceph/pull/59580.
as for [this](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L6035-L6049) block, I think there was a particular edge case for the `current_gen` we were trying to work around. I do think, as Casey said, should return set_cr_done() after writing to error repo there.

[the remark this answers is not in the document]
Comment (clwluvw) [reply src/rgw/driver/rados/rgw_data_sync.cc:6082]: I see, thanks for the explanation. but so we stuck the replication until that generation is proceeding now if I understand correctly.

> as for [this](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L6035-L6049) block, I think there was a particular edge case for the current_gen we were trying to work around. I do think, as Casey said, should return set_cr_done() after writing to error repo there.

If we return `set_cr_done()` then the `gen` would be lost I believe - basically this was my test case. based on your explanation, I guess reverting that PR should be enough, or what was the edge case?

[the remark this answers is not in the document]
Comment (smanjara) [reply src/rgw/driver/rados/rgw_data_sync.cc]: in the comment [here](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_data_sync.cc#L6036), it says the error repo retry was added due to a sync issue when data log entry is missing. I am not exactly sure how that happens.

**Review by smanjara — APPROVED** [review]

**Comment by smanjara** [comment]
passed https://pulpito.ceph.com/smanjara-2024-11-22_22:48:55-rgw:multisite-rgw-multisite-two-zonegroups-distro-default-smithi/
